cvmfs/kvstore.cc

```diff
@@ -0,0 +1,121 @@
+#include <errno.h>
+#include <string.h>
+#include <assert.h>
+#include <limits.h>
+#include "kvstore.h"
+#include "util_concurrency.h"
+
+using namespace std;  // NOLINT
+
+namespace kvstore {
+
+bool MemoryKvStore::GetBuffer(const shash::Any &id, MemoryBuffer *buf) {
+  return Entries.Lookup(id, buf);
+}
+
+bool MemoryKvStore::PopBuffer(const shash::Any &id, MemoryBuffer *buf) {
+  WriteLockGuard guard(rwlock_);
+  if (!Entries.Lookup(id, buf)) return false;
+  used_bytes -= (*buf).size;
+  Entries.Forget(id);
+  return true;
+}
+
+int64_t MemoryKvStore::GetSize(const shash::Any &id) {
+  MemoryBuffer mem;
+  if (Entries.Lookup(id, &mem)) {
+    return mem.size;
+  } else {
+    return -ENOENT;
+  }
+}
+
+int64_t MemoryKvStore::GetRefcount(const shash::Any &id) {
+  MemoryBuffer mem;
+  if (Entries.Lookup(id, &mem)) {
+    return mem.refcount;
+  } else {
+    return -ENOENT;
+  }
+}
+
+bool MemoryKvStore::Ref(const shash::Any &id) {
+  WriteLockGuard guard(rwlock_);
+  MemoryBuffer mem;
+  if (Entries.Lookup(id, &mem)) {
+    assert(mem.refcount < UINT_MAX);
+    ++mem.refcount;
+    Entries.Insert(id, mem);
+    return true;
+  } else {
+    return false;
+  }
+}
+
+bool MemoryKvStore::Unref(const shash::Any &id) {
+  WriteLockGuard guard(rwlock_);
+  MemoryBuffer mem;
+  if (Entries.Lookup(id, &mem) && mem.refcount > 0) {
```

Comment:
Are we sure we silently want to ignore double unrefs?  It probably is an indicator for a misuse, so we might want to assert, no?